File: api/src/routes/mcp.ts
import { Router, Request, Response, NextFunction } from 'express';
import { db, redis } from '../config';
import { requireApiKey, checkRateLimit } from '../middleware/apiKey';
import { queryLogMiddleware } from '../middleware/queryLog';
import { buildErrorEnvelope, ErrorCode, ErrorCodeType } from '../middleware/errors';
import { buildProduct, buildSearchResponse, COUNTRY_CURRENCY, CURRENCY_RATES } from '../lib/response';

const router = Router();

// MCP tools manifest
const TOOLS = [
  {
    name: 'search_products',
    description: 'Search the BuyWhere product catalog by keyword. Returns products from e-commerce platforms across multiple regions (Singapore, US, etc.). Use compact=true for agent-optimized responses with structured_specs, comparison_attributes, and normalized_price_usd fields.',
    inputSchema: {
      type: 'object',
      properties: {
        q: { type: 'string', description: 'Keyword search query' },
        domain: { type: 'string', description: 'Filter by merchant platform (e.g. lazada, shopee, amazon)' },
        region: { type: 'string', description: 'Filter by region (sea, us, eu, au)' },
        country_code: { type: 'string', enum: ['SG', 'US', 'VN', 'TH', 'MY'], description: 'Filter by ISO country code. Also infers default currency for price filters (SG→SGD, US→USD, VN→VND, TH→THB, MY→MYR).' },
        country: { type: 'string', description: 'Alias for country_code (deprecated, use country_code)' },
        min_price: { type: 'number', description: 'Minimum price (in currency inferred from country_code, or SGD by default)' },
        max_price: { type: 'number', description: 'Maximum price (in currency inferred from country_code, or SGD by default)' },
        limit: { type: 'integer', description: 'Number of results (max 100, default 20)', default: 20 },
        offset: { type: 'integer', description: 'Pagination offset', default: 0 },
        compact: { type: 'boolean', description: 'Return agent-optimized compact shape: structured_specs, comparison_attributes, normalized_price_usd. Reduces response size ~40%. Recommended for agent tool-use.', default: false },
        category: { type: 'string', description: 'Filter by product category name (e.g. "Laptops", "Smartphones", "Televisions"). Use to exclude accessories and get actual products.' },
      },
    },
  },
  {
    name: 'get_product',
    description: 'Get a specific product by its ID, including full details and current price.',
    inputSchema: {
      type: 'object',
      required: ['id'],
      properties: {
        id: { type: 'string', description: 'Product UUID' },
      },
    },
  },
  {
    name: 'compare_products',
    description: 'Compare multiple products side-by-side. Returns price, brand, rating, and category for each.',
    inputSchema: {
      type: 'object',
      required: ['ids'],
      properties: {
        ids: {
          type: 'array',
          items: { type: 'string' },
          description: 'Array of product IDs to compare (2-10)',
          minItems: 2,
          maxItems: 10,
        },
      },
    },
  },
  {
    name: 'get_deals',
    description: 'Get discounted products sorted by discount percentage. Returns products with original price and discount percentage. Supports currency, region (sea, us, eu, au) and country (SG, US, VN, MY, ...) filters.',
    inputSchema: {
      type: 'object',
      properties: {
        min_discount: { type: 'number', description: 'Minimum discount percentage (default 10)', default: 10 },
        currency: { type: 'string', description: 'Filter by currency code (SGD, USD, MYR, VND, THB). Defaults to SGD.', default: 'SGD' },
        region: { type: 'string', description: 'Filter by region (sea, us, eu, au)' },
        country_code: { type: 'string', enum: ['SG', 'US', 'VN', 'TH', 'MY'], description: 'Filter by ISO country code. Alias: country.' },
        country: { type: 'string', description: 'Alias for country_code (deprecated, use country_code)' },
        limit: { type: 'integer', description: 'Number of results (max 100, default 20)', default: 20 },
        offset: { type: 'integer', description: 'Pagination offset', default: 0 },
      },
    },
  },
  {
    name: 'list_categories',
    description: 'List top-level product categories available in the BuyWhere catalog.',
    inputSchema: {
      type: 'object',
      properties: {},
    },
  },
  {
    name: 'find_best_price',
    description: 'Use this whenever a user asks about prices, wants to find the cheapest option, or asks "what\'s the best price for X" or "where can I buy X for the lowest price". This finds the best current price across all merchants.',
    inputSchema: {
      type: 'object',
      required: ['product_name'],
      properties: {
        product_name: { type: 'string', description: 'Product name to find best price for (e.g., "iphone 15 pro 256gb", "samsung galaxy s24")' },
        category: { type: 'string', description: 'Category to filter by (e.g., "electronics", "fashion")' },
        country_code: { type: 'string', enum: ['SG', 'MY', 'TH', 'PH', 'VN', 'ID', 'US'], description: 'Country to search in (defaults to SG). Alias: country.' },
        country: { type: 'string', description: 'Alias for country_code (deprecated, use country_code)' },
        region: { type: 'string', enum: ['us', 'sea'], description: 'Region filter - use "us" for United States or "sea" for Southeast Asia' },
      },
    },
  },
];

// Tool handlers
async function handleSearchProducts(args: Record<string, unknown>) {
  const t0 = Date.now();
  const q = (args.q as string) || '';
  const domain = (args.domain as string) || '';
  const region = (args.region as string) || '';
  // country_code is canonical; `country` kept as alias for backward compat
  // Default to SG when no country/region specified (BUY-6598: SG market is primary)
  const rawCountry = (((args.country_code as string) || (args.country as string)) || '').toUpperCase();
  const country = rawCountry || (!region ? 'SG' : '');
  const category = (args.category as string) || '';
  const minPrice = args.min_price != null ? Number(args.min_price) : null;
  const maxPrice = args.max_price != null ? Number(args.max_price) : null;
  const limit = Math.min(Number(args.limit) || 20, 100);
  const offset = Number(args.offset) || 0;
  const compact = args.compact === true;
  const currency = country ? (COUNTRY_CURRENCY[country] || 'SGD') : 'SGD';

  const cacheKey = `fts:${q}:${domain}:${region}:${country}:${category}:${currency}:${minPrice}:${maxPrice}:${limit}:${offset}:${compact ? 'c' : 'f'}`;
  try {
    const cached = await redis.get(cacheKey);
    if (cached) {
      const parsed = JSON.parse(cached);
      if (parsed.results) {
        return { ...parsed, cached: true, response_time_ms: Date.now() - t0 };
      }
    }
  } catch (_) { /* redis miss — proceed */ }

  const conditions: string[] = [];
  const params: unknown[] = [];

  if (q) {
    params.push(q);
    conditions.push(`search_vector @@ plainto_tsquery('english', $${params.length})`);
  }
  if (domain) {
    params.push(domain);
    conditions.push(`source = $${params.length}`);
  }
  if (minPrice != null) {
    params.push(minPrice);
    conditions.push(`price >= $${params.length}`);
  }
  if (maxPrice != null) {
    params.push(maxPrice);
    conditions.push(`price <= $${params.length}`);
  }
  if (region) {
    params.push(region);
    conditions.push(`region = $${params.length}`);
  }
  if (country) {
    params.push(country.toUpperCase());
    conditions.push(`country_code = $${params.length}`);
  }
  if (category) {
    params.push(`%${category}%`);
    conditions.push(`category ILIKE $${params.length}`);
  }

  const where = conditions.length ? `WHERE ${conditions.join(' AND ')}` : '';

  let rows: unknown[];
  let total: number;

  if (q) {
    // Count matching rows to pick query strategy
    const countResult = await db.query(
      `SELECT COUNT(*) FROM products ${where}`,
      params
    );
    total = parseInt(countResult.rows[0].count, 10);

    if (total <= 1000) {
      params.push(limit, offset);
      const result = await db.query(
        `SELECT id, sku AS source, source AS domain, url, title,
                price, currency, image_url, metadata, updated_at,
                region, country_code,
                ts_rank(search_vector, plainto_tsquery('english', $1)) AS rank
         FROM products ${where}
         ORDER BY rank DESC
         LIMIT $${params.length - 1} OFFSET $${params.length}`,
        params
      );
      rows = result.rows;
    } else {
      const CANDIDATE_LIMIT = Math.min((limit + offset) * 10, 5000);
      params.push(CANDIDATE_LIMIT);
      const candidateResult = await db.query(
        `SELECT id, sku AS source, source AS domain, url, title,
                price, currency, image_url, metadata, updated_at,
                region, country_code,
                ts_rank(search_vector, plainto_tsquery('english', $1)) AS rank
         FROM products ${where}
         ORDER BY rank DESC
         LIMIT $${params.length}`,
        params
      );
      rows = candidateResult.rows.slice(offset, offset + limit);
    }
  } else {
    const countResult = await db.query(`SELECT COUNT(*) FROM products ${where}`, params);
    total = parseInt(countResult.rows[0].count, 10);
    params.push(limit, offset);
    const result = await db.query(
      `SELECT id, sku AS source, source AS domain, url, title,
              price, currency, image_url, metadata, updated_at,
              region, country_code
       FROM products ${where}
       ORDER BY updated_at DESC
       LIMIT $${params.length - 1} OFFSET $${params.length}`,
      params
    );
    rows = result.rows;
  }

  const products = (rows as Record<string, unknown>[]).map(r =>
    buildProduct(r, currency, compact)
  );

  const result = buildSearchResponse(
    products, total!, limit, offset, Date.now() - t0, false
  );

  try {
    await redis.set(cacheKey, JSON.stringify(result), 'EX', 60);
  } catch (_) { /* cache write failure is non-fatal */ }

  return result;
}

async function handleGetProduct(args: Record<string, unknown>) {
  const t0 = Date.now();
  const { id } = args;
  let result;
  try {
    result = await db.query(
      `SELECT id, sku AS source, source AS domain, url, title,
              price, currency, image_url, brand, category_path,
              avg_rating AS rating, review_count, metadata, updated_at, region, country_code
       FROM products WHERE id = $1`,
      [id]
    );
  } catch {
    throw { code: -32001, message: 'Product not found' };
  }
  if (!result.rows.length) throw { code: -32001, message: 'Product not found' };
  const product = buildProduct(result.rows[0] as Record<string, unknown>, 'SGD', false);
  return buildSearchResponse([product], 1, 1, 0, Date.now() - t0, false);
}

async function handleCompareProducts(args: Record<string, unknown>) {
  const t0 = Date.now();
  const ids = args.ids as string[];
  if (!ids || ids.length < 2) throw { code: -32602, message: 'Provide at least 2 product IDs' };
  const placeholders = ids.map((_, i) => `$${i + 1}`).join(',');
  const result = await db.query(
    `SELECT id, sku AS source, source AS domain, url, title,
            price, currency, image_url, brand, category_path,
            avg_rating AS rating, review_count, metadata, updated_at, region, country_code
     FROM products WHERE id IN (${placeholders})`,
    ids
  );
  const products = result.rows.map((r: Record<string, unknown>) => buildProduct(r, 'SGD', false));
  return buildSearchResponse(products, products.length, ids.length, 0, Date.now() - t0, false);
}

async function handleGetDeals(args: Record<string, unknown>) {
  const t0 = Date.now();
  const minDiscount = Number(args.min_discount) || 10;
  const currency = ((args.currency as string) || 'SGD').toUpperCase();
  const region = (args.region as string) || '';
  const country = ((args.country_code as string) || (args.country as string) || '').toUpperCase();
  const limit = Math.min(Number(args.limit) || 20, 100);
  const offset = Number(args.offset) || 0;

  const cacheKey = `deals_mcp:${currency}:${minDiscount}:${region}:${country}:${limit}:${offset}`;
  try {
    const cached = await redis.get(cacheKey);
    if (cached) {
      const parsed = JSON.parse(cached);
      if (parsed.results) {
        return { ...parsed, cached: true, response_time_ms: Date.now() - t0 };
      }
    }
  } catch (_) {}

  const conditions: string[] = [
    `currency = $1`,
    `(metadata->>'original_price')::numeric > price`,
    `(metadata->>'original_price')::numeric < price * 100`,
    `price > 0`,
    `(1 - price / NULLIF((metadata->>'original_price')::numeric, 0)) * 100 >= $2`,
  ];
  const params: unknown[] = [currency, minDiscount];

  if (region) {
    params.push(region);
    conditions.push(`region = $${params.length}`);
  }
  if (country) {
    params.push(country.toUpperCase());
    conditions.push(`country_code = $${params.length}`);
  }

  const whereClause = conditions.join(' AND ');

  const [countResult, dataResult] = await Promise.all([
    db.query(
      `SELECT COUNT(*) FROM products WHERE ${whereClause}`,
      params
    ),
    (() => {
      const dataParams = [...params, limit, offset];
      const limitIdx = dataParams.length - 1;
      const offsetIdx = dataParams.length;
      return db.query(
        `SELECT id, sku AS source, source AS domain, url, title,
                price, (metadata->>'original_price')::numeric AS original_price,
                currency, image_url, metadata, updated_at, region, country_code,
                ROUND((1 - price / NULLIF((metadata->>'original_price')::numeric, 0)) * 100) AS discount_pct
         FROM products
         WHERE ${whereClause}
         ORDER BY discount_pct DESC
         LIMIT $${limitIdx} OFFSET $${offsetIdx}`,
        dataParams
      );
    })(),
  ]);

  const products = dataResult.rows.map((r: Record<string, unknown>) =>
    buildProduct(r, currency, false)
  );
  const total = parseInt(countResult.rows[0].count, 10);
  const result = buildSearchResponse(products, total, limit, offset, Date.now() - t0, false);

  redis.set(cacheKey, JSON.stringify(result), 'EX', 60).catch(() => {});

  return result;
}

async function handleListCategories(_args: Record<string, unknown>) {
  const t0 = Date.now();
  const cacheKey = 'categories_mcp:top100';
  try {
    const cached = await redis.get(cacheKey);
    if (cached) {
      const parsed = JSON.parse(cached);
      return { ...parsed, meta: { ...parsed.meta, cached: true, response_time_ms: Date.now() - t0 } };
    }
  } catch (_) {}

  const result = await db.query(
    `SELECT category_path[1] AS slug,
            category_path[1] AS name,
            COUNT(*) AS product_count
     FROM products
     WHERE category_path IS NOT NULL AND array_length(category_path, 1) > 0
     GROUP BY 1
     ORDER BY product_count DESC
     LIMIT 100`
  );
  const data = { data: result.rows, meta: { total: result.rows.length, response_time_ms: Date.now() - t0, cached: false } };
  redis.set(cacheKey, JSON.stringify(data), 'EX', 300).catch(() => {});
  return data;
}

async function handleFindBestPrice(args: Record<string, unknown>) {
  const t0 = Date.now();
  const productName = (args.product_name as string) || '';
  if (!productName) throw { code: -32602, message: 'product_name is required' };

  const country = (((args.country_code as string) || (args.country as string)) || 'SG').toUpperCase();
  const region = (args.region as string) || '';
  const category = (args.category as string) || '';
  const limit = 10;

  const conditions: string[] = ['is_active = true'];
  const params: unknown[] = [];

  params.push(productName);
  conditions.push(`search_vector @@ plainto_tsquery('english', $${params.length})`);

  if (country) {
    params.push(country);
    conditions.push(`country_code = $${params.length}`);
  }
  if (region) {
    params.push(region);
    conditions.push(`region = $${params.length}`);
  }
  if (category) {
    params.push(`%${category}%`);
    conditions.push(`category ILIKE $${params.length}`);
  }

  params.push(limit);
  const where = `WHERE ${conditions.join(' AND ')}`;
  const result = await db.query(
    `SELECT id, title, price, currency, source AS domain, url, image_url,
            country_code,
            ts_rank(search_vector, plainto_tsquery('english', $1)) AS rank
     FROM products ${where}
     ORDER BY price ASC, rank DESC
     LIMIT $${params.length}`,
    params
  );

  const currency = COUNTRY_CURRENCY[country] || 'SGD';
  const toUsd = CURRENCY_RATES[currency] ?? 1;

  const data = result.rows.map((r: Record<string, unknown>) => ({
    id: r.id,
    title: r.title,
    price: { amount: r.price != null ? parseFloat(r.price as string) : null, currency: r.currency || currency },
    normalized_price_usd: r.price != null ? Math.round(Number(r.price) * toUsd * 100) / 100 : null,
    merchant: r.domain as string,
    url: r.url as string,
    image_url: r.image_url as string,
    country_code: r.country_code as string,
  }));

  return {
    best_price: data[0] ?? null,
    alternatives: data.slice(1),
    meta: { total: data.length, country, response_time_ms: Date.now() - t0 },
  };
}

async function dispatchTool(name: string, args: Record<string, unknown>) {
  switch (name) {
    case 'search_products':  return handleSearchProducts(args);
    case 'get_product':      return handleGetProduct(args);
    case 'compare_products': return handleCompareProducts(args);
    case 'get_deals':        return handleGetDeals(args);
    case 'list_categories':  return handleListCategories(args);
    case 'find_best_price':  return handleFindBestPrice(args);
    default:
      throw { code: -32601, message: `Unknown tool: ${name}` };
  }
}

// JSON-RPC 2.0 response helpers
function jsonrpcOk(id: unknown, result: unknown) {
  return { jsonrpc: '2.0', id, result };
}
function jsonrpcErr(id: unknown, code: number, message: string, data?: unknown, envelopeCode?: string) {
  const errorData: Record<string, unknown> = data != null ? { detail: data } : {};
  if (envelopeCode) {
    errorData.envelope = buildErrorEnvelope(envelopeCode as ErrorCodeType, message);
  }
  return { jsonrpc: '2.0', id, error: { code, message, ...(Object.keys(errorData).length ? { data: errorData } : {}) } };
}

// GET /mcp — info endpoint for browser / reviewer verification.
// Returns a JSON descriptor instead of Express's default 404 so registry
// reviewers and DevRel verifiers can confirm the endpoint is live without
// needing to craft a JSON-RPC POST. The actual MCP protocol uses POST only.
router.get('/', (_req: Request, res: Response) => {
  res.json({
    name: 'buywhere-catalog',
    description: 'BuyWhere MCP server. JSON-RPC 2.0 over HTTP POST.',
    protocol: 'mcp',
    protocolVersion: '2024-11-05',
    transport: 'http',
    methods: ['initialize', 'tools/list', 'tools/call'],
    tools: TOOLS.map(t => t.name),
    auth: 'Bearer token — register at https://api.buywhere.ai/v1/auth/register',
    usage: 'POST this URL with a JSON-RPC 2.0 envelope. See https://api.buywhere.ai/docs/guides/mcp',
  });
});

// POST /mcp — public methods (no auth): initialize + tools/list
// Directory scanners (Glama, Smithery) call these without credentials to introspect the server.
router.post('/', async (req: Request, res: Response, next: NextFunction) => {
  const body = req.body;
  if (!body || body.jsonrpc !== '2.0' || !body.method) {
    return next(); // let the authenticated handler return the 400
  }
  const { id, method } = body;
  if (method === 'initialize') {
    return res.json(jsonrpcOk(id, {
      protocolVersion: '2024-11-05',
      capabilities: { tools: {} },
      serverInfo: { name: 'buywhere-catalog', version: '1.0.0' },
    }));
  }
  if (method === 'tools/list') {
    return res.json(jsonrpcOk(id, { tools: TOOLS }));
  }
  return next();
});

// POST /mcp — authenticated methods: tools/call (and any future additions)
router.post('/', requireApiKey, checkRateLimit, queryLogMiddleware('mcp'), async (req: Request, res: Response) => {
  const body = req.body;

  // Validate JSON-RPC envelope
  if (!body || body.jsonrpc !== '2.0' || !body.method) {
    return res.status(400).json(jsonrpcErr(body?.id ?? null, -32600, 'Invalid JSON-RPC request', undefined, ErrorCode.INVALID_JSON));
  }

  const { id, method, params } = body;
  const args = (params && typeof params === 'object' && !Array.isArray(params)) ? params : {};

  try {
    switch (method) {
      case 'tools/call': {
        const toolName = args.name as string;
        const toolArgs = (args.arguments && typeof args.arguments === 'object') ? args.arguments as Record<string, unknown> : {};
        if (!toolName) {
          return res.json(jsonrpcErr(id, -32602, 'Missing tool name'));
        }
        const result = await dispatchTool(toolName, toolArgs);
        return res.json(jsonrpcOk(id, {
          content: [{ type: 'text', text: JSON.stringify(result) }],
        }));
      }

      default:
        return res.json(jsonrpcErr(id, -32601, `Method not found: ${method}`));
    }
  } catch (err: unknown) {
    const e = err as { code?: number; message?: string };
    if (typeof e.code === 'number' && e.message) {
      const envelopeCode = e.code === -32001 ? ErrorCode.NOT_FOUND
        : e.code === -32602 ? ErrorCode.INVALID_PARAMETER
        : ErrorCode.INTERNAL_ERROR;
      return res.json(jsonrpcErr(id, e.code, e.message, undefined, envelopeCode));
    }
    console.error('[mcp] error:', err);
    return res.json(jsonrpcErr(id, -32603, 'Internal error', undefined, ErrorCode.INTERNAL_ERROR));
  }
});

export default router;

File: src/app/routers/mcp.py
import logging
from typing import Any
from contextvars import ContextVar

from fastapi import APIRouter, Depends, Request, Response
from pydantic import BaseModel

from app.auth import get_current_api_key
from app.models.product import ApiKey

_mcp_api_key_var: ContextVar[str | None] = ContextVar("mcp_api_key", default=None)

from mcp.server import Server
from mcp.types import (
    CallToolResult,
    ListToolsResult,
    TextContent,
    Tool,
)

logger = logging.getLogger("mcp-http")

router = APIRouter(prefix="/mcp", tags=["mcp"])

_api_server: Server | None = None


def get_mcp_server() -> Server:
    global _api_server
    if _api_server is None:
        server = Server("buywhere")

        @server.list_tools()
        async def list_tools() -> ListToolsResult:
            return ListToolsResult(
                tools=[
                    Tool(
                        name="search_products",
                        description=(
                            "Search the BuyWhere product catalog by keyword. "
                            "Returns ranked results from Singapore e-commerce platforms "
                            "(Lazada, Shopee, Qoo10, Carousell)."
                        ),
                        inputSchema={
                            "type": "object",
                            "properties": {
                                "query": {"type": "string", "description": "Product search query."},
                                "category": {"type": "string", "description": "Optional category filter."},
                                "min_price": {"type": "number", "description": "Minimum price in SGD."},
                                "max_price": {"type": "number", "description": "Maximum price in SGD."},
                                "source": {
                                    "type": "string",
                                    "description": "Platform filter (lazada_sg, shopee_sg, etc.).",
                                },
                                "limit": {
                                    "type": "integer",
                                    "description": "Max results (default 10, max 50).",
                                    "default": 10,
                                    "minimum": 1,
                                    "maximum": 50,
                                },
                            },
                            "required": ["query"],
                        },
                    ),
                    Tool(
                        name="get_product",
                        description="Retrieve full details for a specific product by its BuyWhere ID.",
                        inputSchema={
                            "type": "object",
                            "properties": {
                                "product_id": {
                                    "type": "integer",
                                    "description": "The BuyWhere product ID.",
                                },
                            },
                            "required": ["product_id"],
                        },
                    ),
                    Tool(
                        name="find_best_price",
                        description=(
                            "Find the single cheapest listing for a product across all Singapore "
                            "e-commerce platforms. Returns the platform, price, and affiliate URL "
                            "for the lowest available price."
                        ),
                        inputSchema={
                            "type": "object",
                            "properties": {
                                "product_name": {
                                    "type": "string",
                                    "description": "Product name or search query.",
                                },
                                "category": {
                                    "type": "string",
                                    "description": "Optional category to narrow the search.",
                                },
                            },
                            "required": ["product_name"],
                        },
                    ),
                    Tool(
                        name="get_deals",
                        description=(
                            "Find products with significant price drops compared to their original "
                            "price. Returns deals sorted by discount percentage with current price, "
                            "original price, and savings."
                        ),
                        inputSchema={
                            "type": "object",
                            "properties": {
                                "category": {
                                    "type": "string",
                                    "description": "Optional category filter (e.g. 'electronics').",
                                },
                                "min_discount_pct": {
                                    "type": "number",
                                    "description": "Minimum discount percentage (default 10).",
                                    "default": 10,
                                    "minimum": 0,
                                    "maximum": 100,
                                },
                                "limit": {
                                    "type": "integer",
                                    "description": "Max results (default 10, max 50).",
                                    "default": 10,
                                    "minimum": 1,
                                    "maximum": 50,
                                },
                            },
                            "required": [],
                        },
                    ),
                ]
            )

        @server.call_tool()
        async def call_tool(name: str, arguments: dict[str, Any]) -> CallToolResult:
            if name == "search_products":
                return await _handle_search_products(arguments)
            if name == "get_product":
                return await _handle_get_product(arguments)
            if name == "find_best_price":
                return await _handle_find_best_price(arguments)
            if name == "get_deals":
                return await _handle_get_deals(arguments)
            return CallToolResult(
                content=[TextContent(type="text", text=f"Unknown tool: {name}")],
                isError=True,
            )

        _api_server = server

    return _api_server


async def _handle_search_products(args: dict[str, Any]) -> CallToolResult:
    from unittest.mock import AsyncMock

    query = str(args.get("query", "")).strip()
    if not query:
        return CallToolResult(
            content=[TextContent(type="text", text="Error: query is required")],
            isError=True,
        )

    params = {"q": query, "limit": min(int(args.get("limit", 10)), 50)}
    for key in ("category", "min_price", "max_price", "source"):
        if args.get(key) is not None:
            params[key] = args[key]

    try:
        data = await _api_get("/v1/products/search", params)
    except Exception as exc:
        logger.exception("search_products API error for %r", query)
        return CallToolResult(
            content=[TextContent(type="text", text=f"Search failed: {exc}")],
            isError=True,
        )

    items = data.get("items", []) if isinstance(data, dict) else []
    if not items:
        return CallToolResult(
            content=[TextContent(type="text", text=f"No products found for: {query}")]
        )

    lines = [f"Found {len(items)} product(s) for **{query}**:\n"]
    for i, p in enumerate(items, 1):
        lines.append(_fmt_product_summary(i, p))
    return CallToolResult(content=[TextContent(type="text", text="\n".join(lines))])


async def _handle_get_product(args: dict[str, Any]) -> CallToolResult:
    product_id = args.get("product_id")
    if not product_id:
        return CallToolResult(
            content=[TextContent(type="text", text="Error: product_id is required")],
            isError=True,
        )
    try:
        data = await _api_get(f"/v1/products/{product_id}")
    except Exception as exc:
        logger.exception("get_product API error for id %r", product_id)
        return CallToolResult(
            content=[TextContent(type="text", text=f"Fetch failed: {exc}")],
            isError=True,
        )
    return CallToolResult(content=[TextContent(type="text", text=_fmt_product_detail(data))])


async def _handle_find_best_price(args: dict[str, Any]) -> CallToolResult:
    product_name = str(args.get("product_name", "")).strip()
    if not product_name:
        return CallToolResult(
            content=[TextContent(type="text", text="Error: product_name is required")],
            isError=True,
        )
    params = {"q": product_name}
    if args.get("category"):
        params["category"] = args["category"]

    try:
        p = await _api_get("/v1/products/best-price", params)
    except Exception as exc:
        logger.exception("find_best_price API error for %r", product_name)
        return CallToolResult(
            content=[TextContent(type="text", text=f"Search failed: {exc}")],
            isError=True,
        )

    if not p or not isinstance(p, dict):
        return CallToolResult(
            content=[TextContent(type="text", text=f"No products found for: {product_name}")]
        )

    price_str = _fmt_price(p.get("price"), p.get("currency", "SGD"))
    affiliate = p.get("affiliate_url") or p.get("buy_url") or ""
    lines = [
        f"## Best Price: {p.get('name', 'Unknown')}",
        f"**Platform:** {p.get('source', 'unknown')}",
        f"**Price:** {price_str}",
        f"**Category:** {p.get('category') or 'N/A'}",
    ]
    if affiliate:
        lines.append(f"**Affiliate URL:** {affiliate}")
    lines.append(f"**Product ID:** {p.get('id', '')}")
    return CallToolResult(content=[TextContent(type="text", text="\n".join(lines))])


async def _handle_get_deals(args: dict[str, Any]) -> CallToolResult:
    min_discount_pct = float(args.get("min_discount_pct", 10))
    limit = min(int(args.get("limit", 10)), 50)
    params = {"min_discount_pct": min_discount_pct, "limit": limit}
    if args.get("category"):
        params["category"] = args["category"]

    try:
        data = await _api_get("/v1/deals", params)
    except Exception as exc:
        logger.exception("get_deals API error")
        return CallToolResult(
            content=[TextContent(type="text", text=f"Deals fetch failed: {exc}")],
            isError=True,
        )

    items = data.get("items", []) if isinstance(data, dict) else []
    if not items:
        return CallToolResult(
            content=[TextContent(type="text", text=f"No deals found with >={min_discount_pct}% discount.")]
        )

    lines = [f"Found {len(items)} deal(s) with >={min_discount_pct}% discount:\n"]
    for i, d in enumerate(items, 1):
        current = _fmt_price(d.get("price"), d.get("currency", "SGD"))
        original = _fmt_price(d.get("original_price"), d.get("currency", "SGD")) if d.get("original_price") else "N/A"
        discount = d.get("discount_pct", 0) or 0
        lines.append(
            f"{i}. **{d.get('name', 'Unknown')}**\n"
            f"   Current: {current} | Was: {original} | Discount: {discount}%\n"
            f"   Platform: {d.get('source', 'unknown')} | ID: {d.get('id', '')}\n"
        )
    return CallToolResult(content=[TextContent(type="text", text="\n".join(lines))])


async def _api_get(path: str, params: dict[str, Any] | None = None) -> Any:
    import httpx
    from app.config import get_settings
    settings = get_settings()
    API_BASE_URL = settings.app_base_url or "http://localhost:8000"

    headers = {"Accept": "application/json"}
    raw_key = _mcp_api_key_var.get()
    if raw_key:
        headers["Authorization"] = f"Bearer {raw_key}"
    async with httpx.AsyncClient(base_url=API_BASE_URL, headers=headers, timeout=10.0) as client:
        resp = await client.get(path, params=params or {})
        resp.raise_for_status()
        return resp.json()


def _fmt_price(price: Any, currency: str = "SGD") -> str:
    if price is None:
        return "N/A"
    try:
        return f"{currency} {float(price):.2f}"
    except (TypeError, ValueError):
        return str(price)


def _fmt_product_summary(index: int, p: dict[str, Any]) -> str:
    name = p.get("name") or p.get("title") or "Unknown"
    price = _fmt_price(p.get("price"), p.get("currency", "SGD"))
    source = p.get("source", "unknown")
    pid = p.get("id", "")
    url = p.get("affiliate_url") or p.get("buy_url") or ""
    url_line = f"\n   URL: {url}" if url else ""
    return f"{index}. **{name}**\n   Price: {price} | Platform: {source}{url_line}\n   ID: {pid}\n"


def _fmt_product_detail(p: dict[str, Any]) -> str:
    if not isinstance(p, dict):
        return str(p)
    lines = [f"## {p.get('name') or 'Product'}"]
    for key, label in [
        ("id", "ID"),
        ("source", "Platform"),
        ("price", "Price"),
        ("currency", "Currency"),
        ("category", "Category"),
        ("affiliate_url", "Affiliate URL"),
        ("buy_url", "Buy URL"),
        ("image_url", "Image"),
    ]:
        val = p.get(key)
        if val is not None:
            lines.append(f"**{label}:** {val}")
    return "\n".join(lines)


class JSONRPCRequest(BaseModel):
    jsonrpc: str = "2.0"
    method: str
    params: dict[str, Any] | None = None
    id: Any = None


class JSONRPCResponse(BaseModel):
    jsonrpc: str = "2.0"
    id: Any
    result: Any | None = None
    error: dict[str, Any] | None = None


@router.get("/health", include_in_schema=False)
async def mcp_health():
    return {"status": "ok", "service": "mcp"}


@router.head("/health", include_in_schema=False)
async def mcp_health_head():
    return Response(status_code=200)


@router.post("/v1/tools/list")
async def list_tools(request: Request, api_key: ApiKey = Depends(get_current_api_key)):
    server = get_mcp_server()
    result = await server.list_tools()
    return JSONRPCResponse(id="pending", result=result)


@router.post("/v1/tools/call")
async def call_tool(
    request: Request,
    body: JSONRPCRequest,
    api_key: ApiKey = Depends(get_current_api_key),
):
    auth_header = request.headers.get("authorization", "")
    raw_key = auth_header.removeprefix("Bearer ").strip()
    _mcp_api_key_var.set(raw_key)
    try:
        server = get_mcp_server()
        result = await server.call_tool(body.method, body.params or {})
        return JSONRPCResponse(id=body.id, result=result)
    except Exception as exc:
        logger.exception("MCP tool call error: %s %s", body.method, exc)
        return JSONRPCResponse(
            id=body.id,
            error={"code": -32603, "message": str(exc)}
        )
    finally:
        _mcp_api_key_var.set(None)